### PR TITLE
SCALRCORE-21436 [go-scalr] Scalr Provider > Provider configuration resource and data source

### PR DESCRIPTION
**Pull request by DayS1eeper**

### CODE REVIEW

**Review by brezitskiy — APPROVED** [review]

**Review by artemvang — APPROVED** [review]

File: provider_configuration.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+		err error,
+	)
+	CreateParameters(
+		ctx context.Context,
+		configurationID string,
+		optionsList *[]ProviderConfigurationParameterCreateOptions,
+	) (
+		created []ProviderConfigurationParameter,
+		err error,
+	)
```
Comment (penja) [inline provider_configuration.go:40]: @DayS1eeper @brezitskiy Навішо там тут ці екшини? На мою думку не варто закладати бізнес лоігку у клієнт який має бути тупим як дверна ручка, це все варто перенести безпосередньо у провайдер.

Comment (DayS1eeper) [reply provider_configuration.go:40]: @penja I thought about it and decided to leave it here. These methods do not implement any business logic (as `syncArguments` from `terraform-provider-scalr` do). `CreateParameters` and `ChangeParameters` only parallelize requests execution.

> client is not right place for business logic
`Module.ReadBySource`, `Workspace.Read` - these methods implement business logic, they are not just a wrapper of API.

Comment (penja) [inline provider_configuration.go:40]: @DayS1eeper I still don't understand why we need this action with difference list of parameters why we don't create and use simple CRUD for  provider_configuration_paramters. It looks like ugly hack until our API has not support multiple request, I have not seen any advantages to add method like this into go clients. If you have question, let's discuss to with @brezitskiy

Comment (penja) [reply provider_configuration.go:40]: @DayS1eeper About parallelize requests execution, we can do this one also on terraform provider layer

File: provider_configuration.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+	ProviderName          string `jsonapi:"attr,provider-name"`
+	ExportShellVariables  bool   `jsonapi:"attr,export-shell-variables"`
+	AwsAccessKey          string `jsonapi:"attr,aws-access-key"`
+	AwsSecretKey          string `jsonapi:"attr,aws-secret-key"`
+	AzurermClientId       string `jsonapi:"attr,azurerm-client-id"`
+	AzurermClientSecret   string `jsonapi:"attr,azurerm-client-secret"`
+	AzurermSubscriptionId string `jsonapi:"attr,azurerm-subscription-id"`
+	AzurermTenantId       string `jsonapi:"attr,azurerm-tenant-id"`
+	GoogleProject         string `jsonapi:"attr,google-project"`
+	GoogleCredentials     string `jsonapi:"attr,google-credentials"`
```
Comment (penja) [inline provider_configuration.go:67]: @DayS1eeper Why we don't add `omitempty` how is will work?

File: provider_configuration_link.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+)
+
+// Compile-time proof of interface implementation.
+var _ ProviderConfigurationLinks = (*providerConfigurationLinks)(nil)
+
+// ProviderConfigurationLinks describes all the provider configurartion link related methods that the Scalr API supports.
+type ProviderConfigurationLinks interface {
+	WorkspaceList(ctx context.Context, workspaceID string, options ProviderConfigurationLinksListOptions) (*ProviderConfigurationLinksList, error)
+	WorkspaceCreate(ctx context.Context, workspaceID string, options ProviderConfigurationWorkspaceLinkCreateOptions) (*ProviderConfigurationLink, error)
+	EnvironmentCreate(ctx context.Context, environmentID string, options ProviderConfigurationEnvironmentLinkCreateOptions) (*ProviderConfigurationLink, error)
```
Comment (penja) [inline provider_configuration_link.go:17]: @DayS1eeper @brezitskiy IMHO is a bad naming we don't create workspace, we link to workspace. Maybe is should be moved to workspace and environment modules???

**Review by penja — CHANGES_REQUESTED** [review]

File: provider_configuration.go (hunk 2 above)
Comment (DayS1eeper) [reply provider_configuration.go:46]: @penja Because it does not affect anything.

File: provider_configuration.go (hunk 1 above)
Comment (DayS1eeper) [reply provider_configuration.go]: @penja moved this to provider

**Review by penja — APPROVED** [review]

**Pull request merged**